File: AgoraChat-ObjectiveC/AgoraChat/Code/Controllers/Chats/Views/ChatCell/ACDPinMessageCell.m
//
//  ACDPinMessageCell.m
//  AgoraChat-Demo
//
//  Created by li xiaoming on 2024/3/12.
//  Copyright © 2024 easemob. All rights reserved.
//

#import "ACDPinMessageCell.h"
#import <chat-uikit/EaseEmojiHelper.h>
#import <SDWebImage/UIImageView+WebCache.h>

@implementation ACDPinMessageCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style
              reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        [self setupSubviews];
    }
    return self;
}

- (void)awakeFromNib {
    [super awakeFromNib];
    // Initialization code
}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
    [super setSelected:selected animated:animated];

    // Configure the view for the selected state
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        [self setupSubviews];
    }
    return self;
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    self.contentView.frame = UIEdgeInsetsInsetRect(self.contentView.frame, UIEdgeInsetsMake(5, 10, 5, 30));
}

- (void)setupSubviews
{
    self.contentView.backgroundColor = [[UIColor alloc] initWithRed:0.96 green:0.96 blue:0.96 alpha:1.0];;
    self.contentView.layer.cornerRadius = 8;
    [self.contentView addSubview:self.titleLabel];
    [self.contentView addSubview:self.dateLabel];
    [self.contentView addSubview:self.messageLabel];
    [self.contentView addSubview:self.image];
    [self.contentView addSubview:self.unpinButton];
    [self.titleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(self.contentView).offset(8);
        make.top.equalTo(self.contentView).offset(5);
        make.height.equalTo(@18);
    }];
    [self.dateLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.contentView).offset(5);
        make.right.equalTo(self.contentView).offset(-8);
    }];
    [self.unpinButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.bottom.equalTo(self.contentView).offset(-5);
        make.right.equalTo(self.contentView).offset(-8);
        make.width.height.equalTo(@20);
    }];
    [self.messageLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.bottom.equalTo(self.contentView).offset(-5);
        make.left.equalTo(self.contentView).offset(8);
        make.right.equalTo(self.unpinButton.mas_left).offset(-10);
    }];
    [self.image mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(self.contentView).offset(8);
        make.bottom.equalTo(self.contentView).offset(-5);
        make.width.equalTo(@80);
        make.height.equalTo(@100);
    }];
    
    
}

- (UILabel *)titleLabel
{
    if (!_titleLabel) {
        _titleLabel = [[UILabel alloc] init];
        _titleLabel.font = [UIFont fontWithName:@"SFPro-Medium" size:10];
        _titleLabel.textColor = [UIColor colorWithRed:0.092 green:0.102 blue:0.108 alpha:1.0];
    }
    return _titleLabel;
}

- (UILabel *)dateLabel
{
    if (!_dateLabel) {
        _dateLabel = [[UILabel alloc] init];
        _dateLabel.font = [UIFont fontWithName:@"SFPro-Regular" size:10];
        _dateLabel.textColor = [UIColor colorWithRed:0.676 green:0.706 blue:0.724 alpha:1.0];
    }
    return _dateLabel;
}

- (UILabel *)messageLabel
{
    if (!_messageLabel) {
        _messageLabel = [[UILabel alloc] init];
        _messageLabel.numberOfLines = 1;
        _messageLabel.font = [UIFont fontWithName:@"SFPro-Regular" size:12];
        _messageLabel.textColor = [UIColor colorWithRed:0.4 green:0.4 blue:0.4 alpha:1.0];
    }
    return _messageLabel;
}

- (UIImageView *)image
{
    if (!_image) {
        _image = [[UIImageView alloc] init];
    }
    return _image;
}

- (UIButton *)unpinButton
{
    if (!_unpinButton) {
        _unpinButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [_unpinButton setImage:[UIImage imageNamed:@"unpin"] forState:UIControlStateNormal];
        [_unpinButton addTarget:self action:@selector(unpinAction) forControlEvents:UIControlEventTouchUpInside];
    }
    return _unpinButton;
}

-(void)unpinAction
{
    if (self.delegate && [self.delegate respondsToSelector:@selector(unpinMessage:)]) {
        [self.delegate unpinMessage:self.model];
    }
}

- (void)updateModelType:(BOOL)isText 
{
    [self.messageLabel removeFromSuperview];
    [self.image removeFromSuperview];
    if (isText) {
        [self.contentView addSubview:self.messageLabel];
        [self.messageLabel mas_makeConstraints:^(MASConstraintMaker *make) {
            make.bottom.equalTo(self.contentView).offset(-5);
            make.left.equalTo(self.contentView).offset(8);
            make.right.equalTo(self.unpinButton.mas_left).offset(-10);
        }];
    } else {
        [self.contentView addSubview:self.image];
        [self.image mas_makeConstraints:^(MASConstraintMaker *make) {
            make.left.equalTo(self.contentView).offset(8);
            make.bottom.equalTo(self.contentView).offset(-5);
            make.width.equalTo(@80);
            make.height.equalTo(@100);
        }];
    }
}

- (void)setModel:(ACDPinMessageModel *)model
{
    _model = model;
    self.titleLabel.text = model.title;
    self.dateLabel.text = model.dateString;
    switch (model.message.body.type) {
        case AgoraChatMessageBodyTypeText:
        {
            AgoraChatTextMessageBody* textBody = (AgoraChatTextMessageBody*)model.message.body;
            self.messageLabel.text = [EaseEmojiHelper convertEmoji:textBody.text];
            [self updateModelType:YES];
        }
            break;
        case AgoraChatMessageBodyTypeFile:
        {
            AgoraChatFileMessageBody* fileBody = (AgoraChatFileMessageBody*)model.message.body;
            self.messageLabel.text = [NSString stringWithFormat:@"[File] %@",fileBody.displayName];
            [self updateModelType:YES];
        }
            break;
        case AgoraChatMessageBodyTypeVoice:
        {
            AgoraChatVoiceMessageBody* voiceBody = (AgoraChatVoiceMessageBody*)model.message.body;
            self.messageLabel.text = [NSString stringWithFormat:@"[Voice] %d",voiceBody.duration];
            [self updateModelType:YES];
        }
            break;
        case AgoraChatMessageBodyTypeCombine:
        {
            AgoraChatCombineMessageBody* combineBody = (AgoraChatCombineMessageBody*)model.message.body;
            self.messageLabel.text = [NSString stringWithFormat:@"[Chat History] %@",combineBody.summary];
            [self updateModelType:YES];
        }
            break;
        case AgoraChatMessageBodyTypeLocation:
        {
            AgoraChatLocationMessageBody* locationBody = (AgoraChatLocationMessageBody*)model.message.body;
            self.messageLabel.text = [NSString stringWithFormat:@"[Location] %@",locationBody.address];
            [self updateModelType:YES];
        }
            break;
        case AgoraChatMessageBodyTypeImage:
        {
            AgoraChatImageMessageBody* imageBody = (AgoraChatImageMessageBody*)model.message.body;
            UIImage * image = [UIImage imageWithContentsOfFile:(imageBody.thumbnailLocalPath.length > 0 ? imageBody.thumbnailLocalPath : imageBody.localPath)];
            if (image)
                self.image.image = image;
            else {
                image = [UIImage imageNamed:@"default"];
                self.image.image = image;
                NSURL* url = [NSURL URLWithString:imageBody.thumbnailRemotePath];
                if (url)
                    [self.image sd_setImageWithURL:url];
            }
            [self updateModelType:NO];
            
        }
            break;
        case AgoraChatMessageBodyTypeVideo:
        {
            AgoraChatVideoMessageBody* videoBody = (AgoraChatVideoMessageBody*)model.message.body;
            UIImage* image = [UIImage imageWithContentsOfFile:videoBody.thumbnailLocalPath];
            if (image) {
                self.image.image = image;
            } else {
                image = [UIImage imageNamed:@"default"];
                self.image.image = image;
                NSURL* url = [NSURL URLWithString:videoBody.thumbnailRemotePath];
                if (url)
                    [self.image sd_setImageWithURL:url];
            }
            self.image.image = image;
            [self updateModelType:NO];
        }
            break;
            
        default:
            break;
    }
}

@end

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Chats/ACDPinMessageCell.swift
//
//  ACDPinMessageCell.swift
//  AgoraChat-Swift
//
//  Created by li xiaoming on 2024/3/15.
//

import UIKit

protocol ACDPinMessageDelegate: AnyObject {
    func unpinMessage(_ message: AgoraChatMessage)
}

class ACDPinMessageCell: UITableViewCell {
    var dateLabel = UILabel()
    var titleLabel = UILabel()
    var messageInfoLabel = UILabel()
    var imgView = UIImageView()
    var unpinButton = UIButton(type: .custom)
    var message: AgoraChatMessage?
    weak var delegate: ACDPinMessageDelegate?

    override func awakeFromNib() {
        super.awakeFromNib()
        // Initialization code
    }
    
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        loadSubViews()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)

        // Configure the view for the selected state
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        self.titleLabel.frame = CGRect(x: 10, y: 5, width: 250, height: 20)
        self.dateLabel.frame = CGRect(x: (self.contentView.frame.width - 150), y: 5, width: 125, height: 20)
        self.messageInfoLabel.frame = CGRect(x: 10, y: 30, width: 200, height: 20)
        self.imgView.frame = CGRect(x: 10, y: 30, width: 80, height: 100)
        self.unpinButton.frame = CGRect(x: (self.contentView.frame.width - 45), y: (self.contentView.frame.height - 30), width: 20, height: 20)
        self.contentView.frame = self.contentView.frame.inset(by: UIEdgeInsets(top: 5, left: 10, bottom: 5, right: 10))
    }
    func loadSubViews() {
        self.contentView.backgroundColor = UIColor(red: 0.96, green: 0.96, blue: 0.96, alpha: 1.0)
        self.contentView.layer.cornerRadius = 8
        titleLabel.font = UIFont(name: "SFPro-Medium", size: 10)
        titleLabel.textColor = UIColor(red: 0.092, green: 0.102, blue: 0.108, alpha: 1.0)
        dateLabel.font = UIFont(name: "SFPro-Regular", size: 10)
        dateLabel.textColor = UIColor(red: 0.676, green: 0.706, blue: 0.724, alpha: 1.0)
        dateLabel.textAlignment = .right
        messageInfoLabel.numberOfLines = 1
        messageInfoLabel.font = UIFont(name: "SFPro-Regular", size: 12)
        messageInfoLabel.textColor = UIColor(red: 0.4, green: 0.4, blue: 0.4, alpha: 1.0)
        unpinButton.setImage(UIImage(named: "unpin"), for: .normal)
        unpinButton.addTarget(self, action: #selector(unpinButtonAction), for: .touchUpInside)
        self.contentView.addSubview(titleLabel)
        self.contentView.addSubview(dateLabel)
        self.contentView.addSubview(messageInfoLabel)
        self.contentView.addSubview(imgView)
        self.contentView.addSubview(unpinButton)
    }
    @objc func unpinButtonAction() {
        if let message = self.message {
            self.delegate?.unpinMessage(self.message!)
        }
    }
    func updateLayout() {
        let text = "\(self.message?.pinnedInfo?.operatorId ?? "") pinned \(self.message?.from ?? "")'s message"
        self.titleLabel.text = text
        
        guard let ts = self.message?.timestamp else { return  }
        let formatter = DateFormatter()
        formatter.dateFormat = "MM-dd HH:mm"
        let date = Date(timeIntervalSince1970: Double(ts) / 1000)
        self.dateLabel.text = formatter.string(from: date)
        
        switch self.message?.body.type {
        case .combine,.text,.location,.voice,.file:
            self.messageInfoLabel.text = message?.pinnedText ?? ""
            self.messageInfoLabel.removeFromSuperview()
            self.imgView.removeFromSuperview()
            self.contentView.addSubview(self.messageInfoLabel)
        case .image:
            if let imageBody = self.message?.body as? AgoraChatImageMessageBody {
                var localPath = imageBody.thumbnailLocalPath
                if localPath?.count == 0 {
                    localPath = imageBody.localPath
                }
                if let localPath = localPath,
                   let image = UIImage.init(contentsOfFile: localPath) {
                    self.imgView.image = image
                } else {
                    self.imgView.setImage(withUrl: imageBody.thumbnailRemotePath, placeholder: "msg_img_broken")
                }
            }
            
            self.messageInfoLabel.removeFromSuperview()
            self.imgView.removeFromSuperview()
            self.contentView.addSubview(self.imgView)
            break
        case .video:
            if let videoBody = self.message?.body as? AgoraChatVideoMessageBody {
                var localPath = videoBody.thumbnailLocalPath
                if localPath?.count == 0 {
                    localPath = videoBody.localPath
                }
                if let localPath = localPath,
                   let image = UIImage.init(contentsOfFile: localPath) {
                    self.imgView.image = image
                } else {
                    self.imgView.setImage(withUrl: videoBody.thumbnailRemotePath, placeholder: "msg_img_broken")
                }
            }
            
            self.messageInfoLabel.removeFromSuperview()
            self.imgView.removeFromSuperview()
            self.contentView.addSubview(self.imgView)
            break
        default:
            break
        }
    }
    func setMessage(_ message: AgoraChatMessage) {
        self.message = message
        updateLayout()
    }
}

extension AgoraChatMessage {
    var pinnedText: String? {
        switch swiftBody {
        case let .text(content: content):
            return content
        case let .file(localPath: _, displayName: dispayName):
            return "[File]\(dispayName)"
        case let .combine(title: _, summary: summary, compatibleText: _, messageIdList: _):
            return "[Chat History]\(summary ?? "")"
        case .voice(localPath: _, displayName: _):
            return "[Voice]"
        case let .location(latitude: _, longitude: _, address: address, buildingName: _):
            return "[Location]\(address ?? "")"
        default:
            return ""
        }
    }
}
